pkg/backoff/exponential.go

```diff
@@ -38,22 +45,51 @@ type ExponentialBackOff struct {
 }
 
 // NewExponentialBackOff creates an istio wrapped ExponentialBackOff.
+// By default, it never stops.
 func NewExponentialBackOff(initFuncs ...func(off *ExponentialBackOff)) BackOff {
-	b := &ExponentialBackOff{}
+	b := ExponentialBackOff{}
 	b.ExponentialBackOff = backoff.NewExponentialBackOff()
+	b.ExponentialBackOff.MaxElapsedTime = 0
 	for _, fn := range initFuncs {
-		fn(b)
+		fn(&b)
 	}
 	b.Reset()
 	return b
 }
 
-const MaxDuration = 1<<63 - 1
-
-func (b *ExponentialBackOff) NextBackOff() time.Duration {
+func (b ExponentialBackOff) NextBackOff() time.Duration {
 	duration := b.ExponentialBackOff.NextBackOff()
 	if duration == b.Stop {
-		return MaxDuration
+		return b.ExponentialBackOff.MaxInterval
 	}
 	return duration
 }
+
+func (b ExponentialBackOff) Reset() {
+	b.ExponentialBackOff.Reset()
+}
+
+// RetryWithContext tries the operation until it does not return error, BackOff stops,
+// or when the context expires, whichever happens first.
+// o is guaranteed to be run at least once.
+// RetryWithContext sleeps the goroutine for the duration returned by BackOff after a
+// failed operation returns.
+func (b ExponentialBackOff) RetryWithContext(ctx context.Context, operation func() error) error {
+	b.Reset()
+	for {
+		err := operation()
+		if err == nil {
+			return nil
+		}
+		next := b.ExponentialBackOff.NextBackOff()
+		if next == b.Stop {
```

Comment:
Wow it can. `b.ExponentialBackOff.NextBackOff()` vs `b.NextBackOff()` are different. That seems confusing..

I wonder if we should stop exposing the internal library at all. We also have the issue that MaxElapsedTime is now the same as MaxInterval with the recent changes. It may make more sense to expose a minimal public interface allowing only MaxInterval,Multiplier, and InitialInterval?

Reply:
Or if not, we should at least comment we are explicitly using the internal NextBackOff not the external one so that we can get b.Stop?